src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,934 @@
+            on_start_pending_bytes: true,
+        }
+    }
+}
+
+struct FlowChecker<E: Engine> {
+    pending_compaction_bytes_soft_limit: u64,
+    pending_compaction_bytes_hard_limit: u64,
+    memtables_threshold: u64,
+    l0_files_threshold: u64,
+
+    cf_checkers: HashMap<String, CFFlowChecker>,
+    throttle_cf: Option<String>,
+    discard_ratio: Arc<AtomicU64>,
+    l0_target_flow: f64,
+    last_target_file: Option<u64>,
+    factor: f64,
+
+    engine: E,
+    recorder: Smoother<30>,
+    limiter: Arc<Limiter>,
+    last_record_time: Instant,
+}
+
+impl<E: Engine> FlowChecker<E> {
+    pub fn new(
+        config: &Config,
+        engine: E,
+        discard_ratio: Arc<AtomicU64>,
+        limiter: Arc<Limiter>,
+    ) -> Self {
+        let mut cf_checkers = map![];
+
+        for cf in engine.kv_engine().cf_names() {
+            cf_checkers.insert(cf.to_owned(), CFFlowChecker::default());
+        }
+
+        Self {
+            pending_compaction_bytes_soft_limit: config.pending_compaction_bytes_soft_limit,
+            pending_compaction_bytes_hard_limit: config.pending_compaction_bytes_hard_limit,
+            memtables_threshold: config.memtables_threshold,
+            l0_files_threshold: config.l0_files_threshold,
+            engine,
+            factor: EMA_FACTOR,
+            discard_ratio,
+            limiter,
+            recorder: Smoother::default(),
+            cf_checkers,
+            throttle_cf: None,
+            l0_target_flow: 0.0,
+            last_target_file: None,
+            last_record_time: Instant::now_coarse(),
+        }
+    }
+
+    fn start(self, rx: Receiver<bool>, flow_info_receiver: Receiver<FlowInfo>) -> JoinHandle<()> {
+        Builder::new()
+            .name(thd_name!("flow-checker"))
+            .spawn(move || {
+                tikv_alloc::add_thread_memory_accessor();
+                let mut checker = self;
+                let mut deadline = std::time::Instant::now();
+                let mut spare_ticks = 0;
+                while rx.try_recv().is_err() {
+                    match flow_info_receiver.recv_deadline(deadline) {
+                        Ok(FlowInfo::L0(cf, l0_bytes)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            checker.check_long_term_l0_files(cf, l0_bytes)
+                        }
+                        Ok(FlowInfo::L0Intra(cf)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            // do nothing for l0 intra compaction, it makes a bigger L0 file
```

Comment:
What if it's a ingested deletion file?

Reply:
It's hard to distinguish it, and it only takes a small part.

Reply:
I think with #7794, this case can happen quite frequently.